core/src/main/java/org/pentaho/di/core/row/value/ValueMetaBase.java

```diff
@@ -4953,6 +4963,242 @@ protected void getOriginalColumnMetadata( ValueMetaInterface v, ResultSetMetaDat
+      int valtype = ValueMetaInterface.TYPE_NONE;
+      boolean isClob = false;
+
+      switch ( originalColumnType ) {
+        case java.sql.Types.CHAR:
+        case java.sql.Types.VARCHAR:
+        case java.sql.Types.NVARCHAR:
+        case java.sql.Types.LONGVARCHAR: // Character Large Object
+          valtype = ValueMetaInterface.TYPE_STRING;
+          length = originalColumnDisplaySize;
+          break;
+
+        case java.sql.Types.CLOB:
+        case java.sql.Types.NCLOB:
+          valtype = ValueMetaInterface.TYPE_STRING;
+          length = DatabaseMeta.CLOB_LENGTH;
+          isClob = true;
+          break;
+
+        case java.sql.Types.BIGINT:
+          // verify Unsigned BIGINT overflow!
+          //
+          if ( originalSigned ) {
+            valtype = ValueMetaInterface.TYPE_INTEGER;
+            precision = 0; // Max 9.223.372.036.854.775.807
+            length = 15;
+          } else {
+            valtype = ValueMetaInterface.TYPE_BIGNUMBER;
+            precision = 0; // Max 18.446.744.073.709.551.615
+            length = 16;
+          }
+          break;
+
+        case java.sql.Types.INTEGER:
+          valtype = ValueMetaInterface.TYPE_INTEGER;
+          precision = 0; // Max 2.147.483.647
+          length = 9;
+          break;
+
+        case java.sql.Types.SMALLINT:
+          valtype = ValueMetaInterface.TYPE_INTEGER;
+          precision = 0; // Max 32.767
+          length = 4;
+          break;
+
+        case java.sql.Types.TINYINT:
+          valtype = ValueMetaInterface.TYPE_INTEGER;
+          precision = 0; // Max 127
+          length = 2;
+          break;
+
+        case java.sql.Types.DECIMAL:
+        case java.sql.Types.DOUBLE:
+        case java.sql.Types.FLOAT:
+        case java.sql.Types.REAL:
+        case java.sql.Types.NUMERIC:
+          valtype = ValueMetaInterface.TYPE_NUMBER;
+          length = originalPrecision;
+          precision = originalScale;
+          if ( length >= 126 ) {
+            length = -1;
+          }
+          if ( precision >= 126 ) {
+            precision = -1;
+          }
+
+          if ( originalColumnType == java.sql.Types.DOUBLE || originalColumnType == java.sql.Types.FLOAT || originalColumnType == java.sql.Types.REAL ) {
+            if ( precision == 0 ) {
+              precision = -1; // precision is obviously incorrect if the type if
+              // Double/Float/Real
+            }
+
+            // If we're dealing with PostgreSQL and double precision types
+            if ( databaseMeta.getDatabaseInterface() instanceof PostgreSQLDatabaseMeta && originalColumnType == java.sql.Types.DOUBLE
+                && precision >= 16 && length >= 16 ) {
+              precision = -1;
+              length = -1;
+            }
+
+            // MySQL: max resolution is double precision floating point (double)
+            // The (12,31) that is given back is not correct
+            if ( databaseMeta.getDatabaseInterface().isMySQLVariant() ) {
```

Comment:
Is there any way we can put this kind of stuff into the specific dialect ? I know this class is littered with these things already, but I'd rather not add more if we can avoid it.